pkg/receive/handler.go

```diff
@@ -877,6 +738,178 @@ func (h *Handler) fanoutForward(pctx context.Context, tenant string, wreqs map[e
+	}
+	return localWrites, remoteWrites, nil
+}
+
+// sendWrites sends the local and remote writes to execute concurrently, controlling them through the provided sync.WaitGroup.
+// The responses from the writes are sent to the responses channel.
+func (h *Handler) sendWrites(
+	ctx context.Context,
+	wg *sync.WaitGroup,
+	requestLogger log.Logger,
+	params remoteWriteParams,
+	localWrites map[endpointReplica]trackedSeries,
+	remoteWrites map[endpointReplica]trackedSeries,
+	responses chan<- writeResponse,
+) {
+	// Do the writes to the local node first. This should be easy and fast.
+	for writeDestination := range localWrites {
+		func(writeDestination endpointReplica) {
+			h.sendLocalWrite(ctx, writeDestination, params.tenant, localWrites[writeDestination], requestLogger, responses)
+		}(writeDestination)
+	}
+
+	// Do the writes to remote nodes. Run them all in parallel.
+	for writeDestination := range remoteWrites {
+		wg.Add(1)
+		go func(writeDestination endpointReplica) {
+			defer wg.Done()
+			h.sendRemoteWrite(ctx, params.tenant, writeDestination, remoteWrites[writeDestination], params.alreadyReplicated, requestLogger, responses)
+		}(writeDestination)
+	}
+}
+
+// sendLocalWrite sends a write request to the local node.
+// The responses are sent to the responses channel.
+func (h *Handler) sendLocalWrite(
+	ctx context.Context,
+	writeDestination endpointReplica,
+	tenant string,
+	trackedSeries trackedSeries,
+	requestLogger log.Logger,
+	responses chan<- writeResponse,
+) {
+	span, tracingCtx := tracing.StartSpan(ctx, "receive_local_tsdb_write")
+	defer span.Finish()
+	span.SetTag("endpoint", writeDestination.endpoint)
+	span.SetTag("replica", writeDestination.replica)
+	err := h.writer.Write(tracingCtx, tenant, &prompb.WriteRequest{
+		Timeseries: trackedSeries.timeSeries,
+	})
+	level.Debug(requestLogger).Log("msg", "wrote to local tsdb", "endpoint", fmt.Sprintf("%v", writeDestination), "err", err)
+	if err != nil {
+		level.Debug(requestLogger).Log("msg", "local tsdb write failed", "err", err.Error())
+		span.SetTag("error", true)
+		span.SetTag("error.msg", err.Error())
+		responses <- newWriteResponse(trackedSeries.seriesIDs, err)
+		return
+	}
+	responses <- newWriteResponse(trackedSeries.seriesIDs, nil)
+}
+
+// sendRemoteWrite sends a write request to the remote node. It takes care of checking wether the endpoint is up or not
+// in the peerGroup, correctly marking them as up or down when appropriate.
+// The responses are sent to the responses channel.
+func (h *Handler) sendRemoteWrite(
+	ctx context.Context,
+	tenant string,
+	endpointReplica endpointReplica,
+	trackedSeries trackedSeries,
+	alreadyReplicated bool,
+	requestLogger log.Logger,
+	responses chan<- writeResponse,
+) {
+	endpoint := endpointReplica.endpoint
+	cl, err := h.peers.get(ctx, endpoint)
+	if err != nil {
+		h.peers.markPeerDown(endpoint)
+		responses <- newWriteResponse(trackedSeries.seriesIDs, err)
+		return
+	}
+	if !h.peers.isPeerUp(endpoint) {
```

Comment:
I'm not sure about it. It seems like `get()` starts doing too many things too many things too and assumes that all the situation in which we need to mark a peer up or down as contained in `get()` -- that's not the case.

Not all the errors in which we mark a peer down will be coming from the `get()` call. There's another possibility that is: the connection is fine, but we get a GRPC code `Unavailable`. Then after the backoff passes, we will try a `get()` again, and the connection will still be fine (nothing to do here), then we successfully remote write into it and **need** to mark that peer as up.

Reply:
Then maybe we can change the names a little bit? I'm not sure if `get` reflects what it does, then, if we only want it to establish a connection and keep a cache for connections. For example:

- `get` -> `getConnection`
- `markPeerDown` -> `markPeerUnavailable`
- `markPeerUp` -> `markPeerAvailable`

I would really like to avoid the up/down names because I think that we normally call something "up" if it's possible to establish a connection with them :smile:

Reply:
I think I can move this specific call to mark peer down after checking the error to inside `get()`, but I can't mark peer up just because I could get the connection. That can only happen if I can also remote write into it.

Reply:
Sounds good to me.

Reply:
Will also move the `isPeerUp` check into `get`, which should allow me to remove it from the interface.